### PR TITLE
feat: ACRCloud language preference with Windows locale auto-detection

### PR DESCRIPTION
**Pull request by aquarion**
## Summary

Returns ACRCloud track title, artist, and album names in the user's preferred language, auto-detected from the Windows locale with an INI override.

Two-mechanism selection layered in `identify_audio()`:
1. **`langs` pass** — picks the localized name from a field's `langs` array using BCP-47 prefix matching (`zh-Hans-CN` → `zh-Hans` → `zh`).
2. **Script-scan pass** — when `langs` doesn't help, scans all `music[]` entries to find one whose primary fields are in the preferred Unicode script (Latin, CJK, Hangul, or Cyrillic).

Language preference is read fresh per call: `[acrcloud] language` in `euterpium.ini` → Windows registry `HKCU\Control Panel\International\LocaleName` → `"en"` fallback.

## Changes

- New `config.get_acrcloud_language()` with lazy `winreg` import (Windows-only)
- New private helpers in `fingerprint.py`: `_dominant_script`, `_preferred_script`, `_pick_lang`, `_pick_field`
- Refactored `identify_audio()` to apply the two-pass selection for title/artist/album
- Documented `language` key in bundled `euterpium.ini`

## Test plan

- [x] 332/332 tests pass (added 30 new tests across `test_config.py` and `test_fingerprint.py`)
- [x] `configured_credentials` fixture patches `get_acrcloud_language` → integration tests are locale-independent
- [x] Edge cases covered: missing `name` keys in `langs` entries, empty `langs` arrays, no-match fallback to `music[0]`, non-Windows registry failure → `"en"`
- [ ] Smoke test on Windows with a real ACRCloud response containing multiple `music[]` entries (e.g. FFXIV OST tracks with both `祖堅正慶` and `Masayoshi Soken` artist entries)

## Spec & Plan

- Spec: \`docs/superpowers/specs/2026-04-26-acrcloud-language-preference-design.md\`
- Plan: \`docs/superpowers/plans/2026-04-26-acrcloud-language-preference.md\`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/fingerprint.py [hunk 1]
```diff
@@ -87,15 +143,38 @@ def identify_audio(wav_bytes: bytes) -> dict | None:
...
+
+        preferred_lang = config.get_acrcloud_language()
+        preferred_script = _preferred_script(preferred_lang)
+
+        def _artist_str(entry: dict) -> str:
+            return ", ".join(a.get("name", "") for a in entry.get("artists", []))
+
+        title = _pick_lang(music.get("title", ""), music.get("langs", []), preferred_lang)
+        if _dominant_script(title) != preferred_script:
+            title = _pick_field(music_list, lambda e: e.get("title", ""), preferred_script)
```
[the remark this answers is not in the document]
Comment (aquarion) [reply app/fingerprint.py]: Fixed in f1782dd. `_pick_lang` now returns `(value, matched)` and the three call sites in `identify_audio` skip the script-scan pass when `matched` is True. Added a unit test (`test_pick_lang_match_takes_priority_over_script`) and an integration test (`test_identify_audio_skips_script_scan_when_langs_matches`) covering the langs-vs-script edge case. Thanks for the catch.

File: app/fingerprint.py [hunk 2]
```diff
@@ -87,15 +143,44 @@ def identify_audio(wav_bytes: bytes) -> dict | None:
...
+        )
+        if not title_matched and _dominant_script(title) != preferred_script:
+            title = _pick_field(music_list, lambda e: e.get("title", ""), preferred_script)
+
+        artist_results = [
+            _pick_lang(a.get("name", ""), a.get("langs", []), preferred_lang)
+            for a in music.get("artists", [])
+        ]
+        artist = ", ".join(value for value, _ in artist_results)
+        all_artists_matched = bool(artist_results) and all(matched for _, matched in artist_results)
```
[the remark this answers is not in the document]
Comment (aquarion) [reply app/fingerprint.py]: Fixed in bdeda4a. `_artist_str` now filters out falsy names with a walrus expression: `name for a in entry.get("artists", []) if (name := a.get("name", ""))`. The `artist_results` join also filters falsy values: `", ".join(value for value, _ in artist_results if value)`. Added a regression test (`test_identify_audio_filters_empty_artist_names`) covering missing-name and empty-name entries.

File: docs/superpowers/specs/2026-04-26-acrcloud-language-preference-design.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+### `fingerprint._pick_lang(primary: str, langs: list[dict], preferred: str) -> str`
+
+Selects the best available name using the `langs` array:
+
+1. Return `primary` immediately if `langs` is empty or `preferred` is empty.
+2. Try exact match: find entry where `code == preferred`.
+3. Progressively strip the trailing subtag and retry: `en-GB` → `en`, `zh-Hans-CN` → `zh-Hans` → `zh`.
+4. Return the matched `name`, or `primary` if nothing matched.
+
```
[the remark this answers is not in the document]
Comment (aquarion) [reply docs/superpowers/specs/2026-04-26-acrcloud-language-preference-design.md]: Updated in bdeda4a. The spec now documents `_pick_lang(...) -> tuple[str, bool]` with the full (value, matched) contract, and the identify_audio section explicitly notes that callers skip the script-scan pass when matched is True.

File: docs/superpowers/plans/2026-04-26-acrcloud-language-preference.md [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+        return primary
+    candidate = preferred
+    while candidate:
+        for entry in langs:
+            if entry.get("code", "").lower() == candidate.lower():
+                return entry["name"]
+        if "-" not in candidate:
+            break
+        candidate = candidate.rsplit("-", 1)[0]
+    return primary
```
[the remark this answers is not in the document]
Comment (aquarion) [reply docs/superpowers/plans/2026-04-26-acrcloud-language-preference.md]: Updated in bdeda4a. Both the Task 4 _pick_lang code snippet (now returning `(value, matched)`) and the corresponding test expectations (now expecting tuples) have been updated. Task 6's identify_audio body also updated to use `title_matched` / `all_artists_matched` / `album_matched` and to filter falsy artist names.

File: app/config.py [hunk 5]
```diff
@@ -155,6 +155,25 @@ def get_acrcloud_access_secret() -> str:
...
+    if explicit:
+        return explicit
+    try:
+        import winreg
+
+        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, r"Control Panel\International") as key:
+            value, _ = winreg.QueryValueEx(key, "LocaleName")
+            stripped = value.strip()
+            return stripped if stripped else "en"
+    except (ImportError, OSError):
```
[the remark this answers is not in the document]
Comment (aquarion) [reply app/config.py:175]: Partially addressed in bdeda4a. Added an explicit `isinstance(value, str)` guard before `.strip()` to handle the documented edge case (registry returning None or non-string), with a new test (`test_get_acrcloud_language_non_string_registry_value_returns_en`). Kept the narrow `except (ImportError, OSError):` catch — the previous review round explicitly narrowed this from `except Exception:` for codebase consistency, and the isinstance guard handles the realistic non-string failure mode without re-broadening.

File: app/fingerprint.py [hunk 6]
```diff
@@ -87,15 +143,44 @@ def identify_audio(wav_bytes: bytes) -> dict | None:
...
+        if not title_matched and _dominant_script(title) != preferred_script:
+            title = _pick_field(music_list, lambda e: e.get("title", ""), preferred_script)
+
+        artist_results = [
+            _pick_lang(a.get("name", ""), a.get("langs", []), preferred_lang)
+            for a in music.get("artists", [])
+        ]
+        artist = ", ".join(value for value, _ in artist_results if value)
+        all_artists_matched = bool(artist_results) and all(matched for _, matched in artist_results)
+        if not all_artists_matched and _dominant_script(artist) != preferred_script:
```
[the remark this answers is not in the document]
Comment (aquarion) [reply app/fingerprint.py]: Fixed in 1cb3759. Switched from `all_artists_matched` to `any_artist_matched` so the script-scan fallback only runs when zero artists matched via langs. Added inline comment in `identify_audio` explaining the rule, plus a regression test (`test_identify_audio_keeps_partial_artist_langs_match`) covering the partial-match scenario you described.

File: docs/superpowers/plans/2026-04-26-acrcloud-language-preference.md [hunk 7]
```diff
@@ -0,0 +1,820 @@
...
+        return explicit
+    try:
+        import winreg
+
+        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, r"Control Panel\International") as key:
+            value, _ = winreg.QueryValueEx(key, "LocaleName")
+            return value.strip() if value.strip() else "en"
+    except Exception:
+        return "en"
+```
```
[the remark this answers is not in the document]
Comment (aquarion) [reply docs/superpowers/plans/2026-04-26-acrcloud-language-preference.md:112]: Updated in 1cb3759. Plan's Task 1 `get_acrcloud_language()` snippet now matches the actual implementation: narrow `(ImportError, OSError)` catch + `isinstance(value, str)` guard, with a brief explanatory note about why it's narrow rather than `except Exception`.

File: docs/superpowers/specs/2026-04-26-acrcloud-language-preference-design.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+In practice, `langs` on artist/album is less common; the more frequent case is multiple `music[]` entries with different primary-language artist names.
+
+## Components
+
+### `config.get_acrcloud_language() -> str`
+
+1. Read `[acrcloud] language` from INI. If non-empty, return it.
+2. Lazy-import `winreg` (startup.py pattern — Windows-only module).
+3. Read `HKCU\Control Panel\International\LocaleName`.
+4. Return the registry value, or `"en"` on any failure.
```
[the remark this answers is not in the document]
Comment (aquarion) [reply docs/superpowers/specs/2026-04-26-acrcloud-language-preference-design.md]: Updated in 1cb3759. Spec now precisely enumerates the failure modes (`ImportError` or `OSError` from registry access, or non-empty-string check on the value) instead of the vague 'any failure' wording.
